kcidb/monitor/spool/__init__.py

```diff
@@ -97,29 +98,49 @@ def post(self, notification, timestamp=None):
             timestamp:      An "aware" datetime.datetime object specifying the
                             notification creation time, or None to use
                             datetime.datetime.now(datetime.timezone.utc).
+            due:            An "aware" datetime.datetime object specifying the
+                            time notification should be sent out,
+                            or None to use "timestamp".
 
         Returns:
-            True, if the notification was posted onto the spool,
+            True, if the notification was posted onto the spool
+            or an existing notification with timestamp past the
+            picked_until time was updated,
             False, if not (it was already there).
         """
         assert isinstance(notification, Notification)
         assert timestamp is None or \
-            isinstance(timestamp, datetime.datetime) and timestamp.tzinfo
+               isinstance(timestamp, datetime.datetime) and timestamp.tzinfo
         if timestamp is None:
             timestamp = datetime.datetime.now(datetime.timezone.utc)
+        assert due is None or \
+               isinstance(due, datetime.datetime) and due.tzinfo
+        if due is None:
+            due = timestamp
+        message = notification.render().as_string(policy=email.policy.SMTPUTF8)
 
-        try:
-            self._get_doc(notification.id).create(dict(
-                created_at=timestamp,
-                # Set to a definitely timed-out time, free for picking
-                picked_until=datetime.datetime.min,
-                message=notification.render().as_string(
-                    policy=email.policy.SMTPUTF8
-                ),
-            ))
-        except Conflict:
-            return False
-        return True
+        @firestore.transactional
+        def create_or_update(transaction, notification, message):
+            snapshot = self._get_doc(notification.id). \
+                get(field_paths=["picked_until"], transaction=transaction)
+            if timestamp > snapshot.get("picked_until"):
+                transaction. \
+                    update(self._get_doc(notification.id), dict(
+                           created_at=timestamp,
+                           message=message))
+                return True
+            try:
+                transaction.set(self._get_doc(notification.id), dict(
+                    created_at=timestamp,
+                    # Set to a definitely timed-out time, free for picking
+                    picked_until=datetime.datetime.min,
+                    message=message,
+                    due=due
+                ))
+            except Conflict:
+                return False
```

Comment:
I think we should be able to handle both creating and updating existing documents with a single call of `set()` here. See the third use case of [set()](https://googleapis.dev/python/firestore/latest/document.html#google.cloud.firestore_v1.document.DocumentReference.set).

Granted we check if the document exists or not, above, and are not proceeding if it exists and is picked, we can do something like this:
```python
transaction.set(
    doc, merge=["message", "due"],
    created_at=timestamp,
    # Set to a definitely timed-out time, free for picking
    picked_until=datetime.datetime.min,
    message=message,
    due=due
)
```
The above should create a document with all the fields, if it doesn't exist, and if it does, only update the `message` and `due` fields.

BTW, could you store the document reference in a variable called `doc` and reuse it, instead of calling `self._get_doc()` every time?